Add documentation for authorisation module. (#3642)
hpi-schul-cloud/schulcloud-server: apps/server/src/modules/authorization/README.md

It include at the moment all avaible permission and is not seperated by concerns and abstaction levels.
The permissions have different implicit scopes like instance, school, or named scope like team and course.

#### (feature flag permissions)

Some others are more feature flags and we want to seperate and move this in future.
Please let not confuse you.

### roles

We have a role collection, where each role has a permissions array that include string based permissions.
The roles can have in the "roles" field other roles to inherit permissions from other roles.
Some of this roles are more technical like user that is only used for inherit.
Some others are scope based with prefix name team*, or course*.

The "real" user roles by name are expert, student, teacher, administrator all are in a school scope and the superhero for the instance.

> In future we want to remove the inherit logic.
> We want to add scopes types to each role.
> Add more technical users for the instance scope.

### entities

The entities are the representation of a single document of the collection, or the type.
It is used for authorisation for now, but should switch to domain object soon.

### domain objects

> They are not really introduced. They should mapped between the repository layer and the domain.

Author: Beides ist richtig. Es muss zwischen den Layern Repo <> Domain gemapped werden. Dabei kann sich das Domain Objekt auch deutlich von der Entität unterscheiden. Zum Beispiel wenn mehre Entitäten zu ein Domain Objekt zusammen geführt werden. Aber sie enthalten trotzdem Werte die etwas von mit der Entität zu tun haben.

You should never implement authorisations on service level, to avoid different authorisation steps.
By calling other micro service for already authorised operations please use API_KEYS.

> Not implemented but coming soon.

## How to use Authorisation Service

### Example 1 - ...

{{ add more }}

### Example 2 - ...

{{ add more }}

### Example 3 - ...

{{ add more }}

## Structure of the Authorisation Components

### feathers-* (legacy/deprecated)

It exist a adapter to call featherJS endpoints that solve authorisations.
> This service is only used in news and should not use on any other place.
> We want to remove it completly.

### Authorisation Module

The authorsation module is the core of authorisation. They collect all needed informations and handle it for you behind a small interface. It export the authoriation service that can be used in your use case over injections.

### Reference.loader

For situations where only the id and the domain object (string) type is knowen, it is possible to use the *ByReferences methodes.
They load the reference directly.
> Please keep in mind that it can be have impact of the performance if you use it wrongly.

> > We do it as seperate method to avoid using in areas where the domain object should be exist, but it is easyer to only pass the id.

Reviewer: ```suggestion
> We keep it as a seperate method to avoid the usage in areas where the domain object should exist, because we see the risk that a developer could be tempted by the ease of only passing the id.
```

### shared/domain/rules/*

The location to add new rules for entities / domain objects.

#### permission-context.builder

> We export a permission context builder for prepare the parameter for the authorisation service called "authorisation context".

Reviewer: ```suggestion
We export a permission context builder to prepare the parameter for the authorisation service called "authorisation context".
```

This is optional and not required.

> But it enable us to change easy the structure of the authorisation context without touching many different places.

Reviewer: ```suggestion
But it enables us to easily change the structure of the authorisation context without touching many different places.
```

### shared/domain/interface/*

#### rolename.enum

> A enum that hold all avaible role names.

Reviewer: ```suggestion
An enum that holds all avaible role names.
```

#### permission.enum

> A enum that hold all avaible permission names. But mixing also all domain scopes atm.

Reviewer: ```suggestion
An enum that holds all avaible permission names, however it's mixing all domain scopes atm.
```

## Working other Internal MicroServices

> Example FileService

{{ add more }}

## Legacy Tech Stack FeatherJS Hooks

{{ add more }}

## Desired Changes in Future

> Some small steps are done. But many open steps still exist.

Reviewer: ```suggestion
Some small steps are done. But many next steps still exist.
```

> They follow our generell target

Reviewer: ```suggestion
They follow our general target.
```

### Next Steps

> 1. implementation of Scope Based Permissions as generell solution instead of User Permissions that has only implicit school scopes for now

Reviewer: ```suggestion
1. Implementation of Scope Based Permissions as general solution instead of User Permissions that only have implicit school scopes for now
```

    Remove populate logic in reference loader
    Solve eager loading in coursegroups

> 2. introdurce system users and api-keys ..it is also needed to handle superhero permissions without overhead in future

Reviewer: ```suggestion
2. Introduce system users and api-keys ..it is also needed to handle superhero permissions without overhead in future
```

>     Think about hasPermission checks from rules to a generall place.

Reviewer: ```suggestion
    Move hasPermission checks from rules to a more generic place.
```

    Remove jwt decorator and cleanup copy logic.
    Move authorisation-context.builder to authorisation module.

> 3. roles without inheritens roles to solve permissions, we want to write it explizit into the collection

Reviewer: ```suggestion
3. Remove inheritance from roles, because we want to write it explizitly into the collection
```

    Moving role api endpoints to nestjs.

>     Fixing of dashboard to do handle roles in the right way as superhero.

Reviewer: ```suggestion
    Fixing of dashboard to handle roles in the right way as superhero.
```

4. switching entity based authorisation to domain objects based in steps
5. cleanup of feature flags from user permissions

>     move exiting feature flags to rules on places where it make sense

Reviewer: ```suggestion
    add existing feature flags to rules on places where it make sense
```

> 6. introduce instance as scope to come to a one implemenation that handle all scopes/rules/permissions/user types in the same way

Reviewer: ```suggestion
6. introduce instance as a scope to have an implemenation that handles all scopes/rules/permissions/user types in the same way
```

### Refactoring Todos

> - Task module need full used authorisation service.

Reviewer: ```suggestion
- Task module should fully use authorisation service.
```

> - News module need the introduce of the authorisation service.

Reviewer: ```suggestion
- News module should start to use authorisation service.
```

### Is Needed

> - we can introduce a new layer called "policy" that can combine different rules (any of them has his one matching strategy) for a single domain object between authorisation and rule

Reviewer: ```suggestion
- we can introduce a new layer called "policy" that combines different rules (any of them has their own matching strategy) for a single domain object between authorisation and rule
```

to reduce complexity in a single rule

> - we can switch that rules register it self by the authorisation service

Reviewer: ```suggestion
- we can switch to a behaviour where rules register themself at the authorisation service than.
```

The location to add new rules for entities / domain objects.

#### permission-context.builder

We export a permission context builder to prepare the parameter for the authorisation service called "authorisation context".
This is optional and not required.
But it enables us to easily change the structure of the authorisation context without touching many different places.

### shared/domain/interface/*

#### rolename.enum

An enum that holds all avaible role names.

#### permission.enum

A enum that hold all avaible permission names, however it's mixing all domain scopes atm.

## Working other Internal MicroServices

> Example FilesStorageService

We have the files storage service application that are a bundle of modules of this repository.
The application is startet as additional micro service.
It exists the need that the server application can call the file service.
We add a files storage client module to the server.
This module export a service to call with the file service.

For the communication it used http calls, but it is switching to RabbitMQ soon.
Every operation must already authorised in the UC of the server. It exist no need to do it again in files storage service.

> For this we want that the consumer of the RabbitMQ call the files storage service, services directly without authorisation.

Reviewer: ```suggestion
For this reason, we want the consumer of the RabbitMQ item to call the files storage service directly without authorisation.
```

## Legacy Tech Stack FeatherJS Hooks

In featherJS all the authorisation is done in hooks. Mostly before hooks and sometimes in after hooks.
Before and after means before, or after the database operation. For self writen services before, or after the call of the operation that should be executed.

> They work similar to express middleware and bring is own request context.

Reviewer: ```suggestion
They work similar to express middleware and bring their own request context.
```

> It exists hooks can be for all http(s) calls, or for specific type based on CRUD operations.

Reviewer: ```suggestion
It exists hooks that can be used for all http(s) calls, or for specific type based on CRUD operations.
```

Additional it exist the find operations that is a http(s) GET without ID of a specific element.

> Each function that add to the hooks will executed in order. Hooks for all methods first, then hooks for specific methodes.

Reviewer: ```suggestion
Each function that adds to the hooks will be executed in order. Hooks for all methods first, then hooks for specific methodes.
```

> Each hooks exist for a featherJS service that exposed directly the api endpoints. Additional it exist a global hook pattern for the whole application.

Reviewer: ```suggestion
Each hook exists for a featherJS service that exposes api endpoints directly. Additional it exists a global hook pattern for the whole application.
```

Example: <https://github.com/hpi-schul-cloud/schulcloud-server/blob/main/src/services/lesson/hooks/index.js#L232>

## Desired Changes in Future

Some small steps are done. But many next steps still exist.
They follow our general target.

### Next Steps

1. Implementation of Scope Based Permissions as generell solution instead of User Permissions that has only implicit school scopes for now.
    Remove populate logic in reference loader.
    Solve eager loading in coursegroups.
2. Introduce RabbitMQ. Splitting Service(logic) from UC, that we can call services over the consumer for internal communication between micro services of already authorised operations.
    Think about: Move hasPermission checks from rules to a more generic place.
    Remove jwt decorator and cleanup copy logic.
    Move authorisation-context.builder to authorisation module.

> 3. Remove inheritance from roles, because we want to write it explizitly into the collection documents.

Reviewer: ```suggestion
3. Remove inheritance from roles, because we want to write it explicitly into the collection documents.
```
